### PR TITLE
[InstCombine] Propagate exact flags in shift-combine transforms

### PR DESCRIPTION
**Pull request by AZero13**
There were a couple of places where we could propagate exact flags but did not.
This patch addresses some of those places.

Alive 2 Proofs:
https://alive2.llvm.org/ce/z/vmoZrX
https://alive2.llvm.org/ce/z/9zxKKA
https://alive2.llvm.org/ce/z/HJebVu
https://alive2.llvm.org/ce/z/96ez9n

### CODE REVIEW

**Comment by XChy** [comment]
Could you please please add alive2 proof for it?

**Comment by nikic** [comment]
@AtariDreams Your contribution behavior continues to be unacceptable, to the point that I am *this close* to requesting a ban from the LLVM organization for you.

Please, do not submit pull requests until you both have adequate test coverage and proofs. See https://llvm.org/docs/InstCombineContributorGuide.html for our contribution guidelines. If you want to submit a PR without these things, submit it as a draft pull request, and only mark it as ready for review when it is actually ready.

**Comment by AtariDreams** [comment]
> Could you please please add alive2 proof for it?

https://alive2.llvm.org/ce/z/vmoZrX

**Comment by AtariDreams** [comment]
> Could you please please add alive2 proof for it?

https://alive2.llvm.org/ce/z/kspgi5

> @AtariDreams Your contribution behavior continues to be unacceptable, to the point that I am _this close_ to requesting a ban from the LLVM organization for you.
> 
> Please, do not submit pull requests until you both have adequate test coverage and proofs. See https://llvm.org/docs/InstCombineContributorGuide.html for our contribution guidelines. If you want to submit a PR without these things, submit it as a draft pull request, and only mark it as ready for review when it is actually ready.

I apologize. I will ensure moving forward to ensure all PRs have proofs and tests.

**Comment by goldsteinn** [comment]
NB: you duplicate your message in commit 2

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1406]: Think you can just `cast<Instruction>(Op0)->isExact()` and keep the original match logic.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1408]: Which proof is assosiated w/ this one?
Just a tip, you can put multiple proofs in the same alive2 by matching postfixes. For example: https://alive2.llvm.org/ce/z/P6HRvw

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1408]: Looks like: https://alive2.llvm.org/ce/z/HJebVu

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1406]: Done!

**Comment by goldsteinn** [comment]
LGTM. Please wait for one more approval before pushing.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by dtcxzyw** [comment]
> Failed Tests (1):
  LLVM :: Transforms/InstCombine/shift.ll

Please update the test.

**Comment by AtariDreams** [comment]
> > Failed Tests (1):
> > LLVM :: Transforms/InstCombine/shift.ll
> 
> Please update the test.

Done!

File: llvm/test/Transforms/InstCombine/shift.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift.ll:462]: Also add a variant where the exact is only on the last shift, rather than only on the first.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1406]: Looks like this is not resolved?

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/lshr.ll:216]: This test (and the next two) are useless in that the pattern only works if both shift amounts are constants. This does not actually trigger the fold.

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/lshr.ll:248]: Okay, so it looks like the exact flag doesn't actually get preserved in this case. Do you know why? Presumably there is some other fold that handles this earlier already.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1406]: > Looks like this is not resolved?

I believe it is now.

**Comment by AtariDreams** [comment]
I will just remove the test for this particular case because for that particular transform there needs to be some more things done.

**Comment by dtcxzyw** [comment]
> Failed Tests (2):
  LLVM :: Transforms/InstCombine/lshr.ll
  LLVM :: Transforms/InstCombine/shift.ll

**Comment by AtariDreams** [comment]
Addressed @dtcxzyw !

**Comment by nikic** [comment]
The test failures in CI are still there:

```
Failed Tests (2):
  LLVM :: Transforms/InstCombine/lshr.ll
  LLVM :: Transforms/InstCombine/shift.ll
```

It would be good to split this PR into separate PRs that handle one fold each. Otherwise it is hard to check whether all cases are covered. Because the tests are not pre-committed, it's also hard to see which of the changes are necessary. E.g. I think the explicit trunc flag preservation may not be needed, as it will get inferred anyway? See https://llvm.godbolt.org/z/MWefaxabj.

**Pull request closed without merging**
